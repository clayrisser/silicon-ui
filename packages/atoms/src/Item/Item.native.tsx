import React, { FC } from 'react';
import { NativeBase, Item as NativeBaseItem } from 'native-base';
import {
  background,
  border,
  compose,
  layout,
  position,
  shadow,
  space
} from 'styled-system';
import ItemContext from '../contexts/Item';
import { createStyled } from '../styled';
import {
  StyledItemProps,
  ItemProps,
  antiForwardItemPropsKeys,
  splitProps
} from './itemProps';

const StyledNativeBaseItem = createStyled<StyledItemProps, NativeBase.Item>(
  NativeBaseItem,
  {
    forwardPropsBlacklist: antiForwardItemPropsKeys
  }
)(compose(background, border, layout, position, shadow, space));

const Item: FC<ItemProps> = (props: ItemProps) => {
  const { styledItemProps, customItemProps, nativeItemProps } = splitProps(
    props
  );

  return (
    <ItemContext.Provider value={{ hasItemParent: true }}>
      <StyledNativeBaseItem
        {...styledItemProps}
        {...nativeItemProps}
        {...customItemProps}
      />
    </ItemContext.Provider>
  );
};

Item.defaultProps = {};

export default Item;
