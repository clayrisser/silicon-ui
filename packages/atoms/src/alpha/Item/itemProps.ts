import { NativeBase } from 'native-base';
import { ReactNode, DetailedHTMLProps, HTMLAttributes } from 'react';
import { Theme } from 'theme-ui';
import {
  BackgroundColorProps,
  BackgroundProps,
  BorderProps,
  LayoutProps,
  OpacityProps,
  PositionProps,
  ShadowProps,
  SpaceProps
} from 'styled-system';

export type DetailedHTMLDivProps = DetailedHTMLProps<
  HTMLAttributes<HTMLDivElement>,
  HTMLDivElement
>;

export interface NativeItemProps extends NativeBase.Item {}

export interface CustomItemProps {
  children?: ReactNode;
  theme?: Theme;
}

export interface StyledItemProps
  extends BackgroundColorProps,
    BackgroundProps,
    BorderProps,
    LayoutProps,
    OpacityProps,
    PositionProps,
    ShadowProps,
    SpaceProps {}

export interface ItemProps
  extends NativeItemProps,
    CustomItemProps,
    StyledItemProps {}

export const customItemPropsKeys = new Set(['theme']);

export const nativeItemPropKeys = new Set([
  'bordered',
  'disabled',
  'error',
  'fixedLabel',
  'floatingLabel',
  'inlineLabel',
  'last',
  'onPress',
  'picker',
  'placeholder',
  'placeholderLabel',
  'regular',
  'rounded',
  'secureTextEntry',
  'stackedLabel',
  'success',
  'underline'
]);

export const antiForwardItemPropsKeys = new Set<string>([]);

export interface SplitProps {
  customItemProps: CustomItemProps;
  nativeItemProps: NativeItemProps;
  styledItemProps: StyledItemProps;
}

export function splitProps(props: ItemProps): SplitProps {
  const styledItemProps: { [key: string]: any } = {};
  const customItemProps: { [key: string]: any } = {};
  const nativeItemProps: { [key: string]: any } = {};
  Object.entries({ ...props }).forEach(([key, prop]: [string, any]) => {
    if (customItemPropsKeys.has(key)) {
      customItemProps[key] = prop;
    } else if (nativeItemPropKeys.has(key)) {
      nativeItemProps[key] = prop;
    } else {
      styledItemProps[key] = prop;
    }
  });
  return {
    customItemProps,
    nativeItemProps,
    styledItemProps
  };
}
