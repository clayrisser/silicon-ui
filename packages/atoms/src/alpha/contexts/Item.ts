import { createContext } from 'react';

export interface ItemContextValue {
  hasItemParent: boolean;
}

export default createContext<ItemContextValue>({
  hasItemParent: false
});
