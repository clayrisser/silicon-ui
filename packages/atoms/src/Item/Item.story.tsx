import React from 'react';
import { withKnobs } from '@storybook/addon-knobs';
import Item from './Item';
import Wrapper from '../../storybook/Wrapper';
import storiesOf from '../../storybook/storiesOf';
import withThemeProvider from '../../storybook/withThemeProvider';
import docs from './docs';

storiesOf('Item', module)
  .addDecorator(withKnobs)
  .addDecorator(withThemeProvider)
  .addParameters({
    docs: { page: docs }
    // jest: ['Item.spec.tsx']
  })
  .add('with knobs', () => (
    <Wrapper>
      <Item />
    </Wrapper>
  ));
