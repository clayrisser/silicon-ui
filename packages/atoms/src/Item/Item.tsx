import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import {
  background,
  border,
  compose,
  layout,
  position,
  shadow,
  space
} from 'styled-system';
import styled, { StyledComponent } from '@emotion/styled';
import ItemContext from '../contexts/Item';
import { ItemProps, DetailedHTMLDivProps, splitProps } from './itemProps';

const HTMLDiv: StyledComponent<
  DetailedHTMLDivProps,
  ItemProps,
  object
> = styled.div(compose(background, border, layout, position, shadow, space));

const Item: FC<ItemProps> = (props: ItemProps) => {
  const { styledItemProps, customItemProps } = splitProps(props);

  function handleClick(_e: any) {
    if (props.onPress) props.onPress();
  }

  return (
    <ItemContext.Provider value={{ hasItemParent: true }}>
      <HTMLDiv
        {...(styledItemProps as any)}
        {...customItemProps}
        onClick={handleClick}
      />
    </ItemContext.Provider>
  );
};

Item.defaultProps = {};

export default Item;
