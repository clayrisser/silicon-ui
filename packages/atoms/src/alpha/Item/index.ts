import Item from './Item';
import { ItemProps } from './itemProps';

export { ItemProps };
export default Item;
