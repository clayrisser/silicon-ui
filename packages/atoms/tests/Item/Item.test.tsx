import React from 'react';
import renderer from 'react-test-renderer';
// import { render, fireEvent } from '@testing-library/react';
import Item from '../../src/Item';

describe('<Item />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Item />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
