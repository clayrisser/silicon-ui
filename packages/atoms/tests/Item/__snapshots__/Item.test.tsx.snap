// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> renders correctly 1`] = `
<div
  className="css-0"
  onClick={[Function]}
/>
`;
